include/mqtt_pal.h

```diff
@@ -85,7 +85,13 @@ extern "C" {
     typedef time_t mqtt_pal_time_t;
     typedef pthread_mutex_t mqtt_pal_mutex_t;
 
-    #define MQTT_PAL_MUTEX_INIT(mtx_ptr) pthread_mutex_init(mtx_ptr, NULL)
+    #define MQTT_PAL_MUTEX_INIT(mtx_ptr) { \
+        pthread_mutexattr_t attr; \
```

Comment:
My concern here is `attr` is declared in the scope of `mqtt_init()` or `mqtt_reconnect()`. I think `pthread_mutexattr_t
` might need to come from the user's program's scope (i.e., the main function's scope, or the same scope as `mqtt_client`'s definition).

What about if the POSIX PAL defines a function like `mqtt_pal_init_pthread_mutex_recursive(struct mqtt_client
*, pthread_mutexattr_t*)` that:
1. Destroys the mutex created by `mqtt_init()` (i.e., the one initialized [here](https://github.com/LiamBindle/MQTT-C/blob/e44bef9aeacb38f7ad4f0b583cdc95428ced1807/examples/simple_publisher.c#L78))
2. Creates a new mutex with `&client->mutex` and the `pthread_mutexattr_t*` passed from the user?

Reply:
An alternative would be changing the `MQTT_PAL_MUTEX_INIT()` interface. It could take two arguments: a `mqtt_pal_mutex_t*` and a `mqtt_pal_mutex_attr_t*`. All the PAL's would need to add the new mutex attr type, but their `MQTT_PAL_MUTEX_INIT` wouldn't use it, so it could be assigned a dummy type like `void*`.

Reply:
Oh, a third option came to mind. What about

```c
struct mqtt_pal_mutex_t {
    pthread_mutex_t  mutex;
    pthread_mutexattr_t attr;
};

The `MQTT_PAL_MUTEX_INIT` is passed a pointer to `mqtt_pal_mutex_t`, so `MQTT_PAL_MUTEX_INIT` could then be
```c
 #define MQTT_PAL_MUTEX_INIT(mtx_ptr) { \
        pthread_mutexattr_t attr; \
        pthread_mutexattr_init(&mtx_ptr->attr); \
        pthread_mutexattr_settype(&mtx_ptr->attr, PTHREAD_MUTEX_RECURSIVE); \
        pthread_mutex_init(&mtx_ptr->mutex, &mtx_ptr->attr); \
    }
```

This is the cleanest solution IMO.

Reply:
That last option does look like a very clean solution. 

Is it worth including an option on whether to use recursive mutexes at all?

```
#ifndef MQTT_USE_MUTEX_RECURSIVE
#define MQTT_PAL_MUTEX_INIT(mtx_ptr) pthread_mutex_init(mtx_ptr, NULL)
#else
#define MQTT_PAL_MUTEX_INIT(mtx_ptr) { \
        pthread_mutexattr_init(&mtx_ptr->attr); \
        pthread_mutexattr_settype(&mtx_ptr->attr, PTHREAD_MUTEX_RECURSIVE); \
        pthread_mutex_init(&mtx_ptr->mutex, &mtx_ptr->attr); \
    }
#endif
```

Or alternatively the flag could be the other way round, e.g. MQTT_USE_MUTEX_DEFAULT, if the recursive mutex is the default. (Possibly a better name for the flag is needed if the default is **not** going to be PTHREAD_MUTEX_DEFAULT.)

Reply:
I don't see a reason to not fully switch to recursive mutexes for UNIX/APPLE. Do you have a concern about it? 

I supose it could lead to confusion if people share their UNIX/APPLE message callback with a different platform. That said, I think this could be considered a "feature" of the UNIX/APPLE PAL, but not something MQTT-C supports in general.

Reply:
No objections from me. It was just my own relative lack of experience with these kind of mutexes.